### PR TITLE
fix: Eliminate from_address_hash == #{address_hash} clause for transactions query in case of smart-contracts

### PR DESCRIPTION
**Pull request by vbaranov**
## Motivation

Adding `from_address_hash == #{address_hash}` on `transactions` table is redundant in case of smart-contracts. In some cases it can lead to long query execution.

## Changelog

Before sending query to the DB we should properly propose where clauses: we should eliminate `from_address_hash == #{address_hash}` where clause for smart-contracts and leave it as is for EOAs. Since smart-contracts cannot be originators of transactions, there are no records satisfying this where clause in case of smart-contract address. Leaving this clause for smart-contracts leads to sequential scan of the whole token transfers table, which leads to long execution time of the original query.

## Checklist for your Pull Request (PR)

  - [ ] I added an entry to `CHANGELOG.md` with this PR
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [ ] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

File: apps/explorer/lib/explorer/etherscan/logs.ex [hunk 1]
```diff
@@ -165,6 +161,19 @@ defmodule Explorer.Etherscan.Logs do
           union: ^internal_transaction_log_query
         )
 
+      dynamic =
+        dynamic(
+          [transaction],
+          ^Transaction.where_transactions_to_from(address_hash) and
+            transaction.block_number >= ^prepared_filter.from_block and
+            transaction.block_number <= ^prepared_filter.to_block
```
Comment (nikitosing) [inline apps/explorer/lib/explorer/etherscan/logs.ex:169]: why did you add additional condition?

File: apps/explorer/lib/explorer/graphql.ex [hunk 2]
```diff
@@ -28,9 +29,10 @@ defmodule Explorer.GraphQL do
   """
   @spec address_to_transactions_query(Hash.Address.t(), :desc | :asc) :: Ecto.Query.t()
   def address_to_transactions_query(address_hash, order) do
+    dynamic = dynamic(^Transaction.where_transactions_to_from(address_hash))
```
Comment (nikitosing) [inline apps/explorer/lib/explorer/graphql.ex:32]: Do we need additionally wrap it with dynamic? As I see the result of the `Transaction .where_transactions_to_from` is `dynamic`

File: apps/explorer/lib/explorer/chain/address/counters.ex [hunk 3]
```diff
@@ -128,10 +128,10 @@ defmodule Explorer.Chain.Address.Counters do
   end
 
   def address_hash_to_transaction_count_query(address_hash) do
-    from(
-      transaction in Transaction,
-      where: transaction.to_address_hash == ^address_hash or transaction.from_address_hash == ^address_hash
-    )
+    dynamic = dynamic(^Transaction.where_transactions_to_from(address_hash))
```
Comment (nikitosing) [inline apps/explorer/lib/explorer/chain/address/counters.ex:131]: Do we need additionally wrap it with dynamic? As I see the result of the `Transaction .where_transactions_to_from` is `dynamic`

File: apps/explorer/lib/explorer/etherscan/logs.ex (hunk 1 above)
Comment (k1rill-fedoseev) [inline apps/explorer/lib/explorer/etherscan/logs.ex:169]: Seems like this will miss a filter for block range in the `transaction.created_contract_address_hash == ^address_hash` case?

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 4]
```diff
@@ -1749,4 +1749,21 @@ defmodule Explorer.Chain.Transaction do
         |> Decimal.min(max_fee_per_gas |> Wei.sub(base_fee_per_gas) |> Wei.to(:wei))
         |> Wei.from(:wei)
   end
+
+  @doc """
+  Dynamically adds to/from for `transactions` query based on whether the target address EOA or smart-contract
+  """
+  @spec where_transactions_to_from(Hash.Address.t()) :: any()
+  def where_transactions_to_from(address_hash) do
```
Comment (k1rill-fedoseev) [inline apps/explorer/lib/explorer/chain/transaction.ex:1757]: Let's mark somewhere that this will have to be rolled back if something like [EIP-5003](https://eips.ethereum.org/EIPS/eip-5003) is included in the roadmap. There is a chance it might come within next 2 planned hardforks.

File: apps/explorer/lib/explorer/etherscan/logs.ex (hunk 1 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/etherscan/logs.ex:169]: Since I added `or_where`, I need to copy this filter to the "or" condition. Otherwise, filtering for block ranges won't work together with to/from filter.

File: apps/explorer/lib/explorer/graphql.ex (hunk 2 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/graphql.ex]: Processed in https://github.com/blockscout/blockscout/pull/9469/commits/b00c0b46fd7a9ff0483decdb6222b7cf6f4b14ac

File: apps/explorer/lib/explorer/chain/address/counters.ex (hunk 3 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/chain/address/counters.ex]: Processed in https://github.com/blockscout/blockscout/pull/9469/commits/b00c0b46fd7a9ff0483decdb6222b7cf6f4b14ac

File: apps/explorer/lib/explorer/chain/transaction.ex (hunk 4 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/chain/transaction.ex:1758]: Processed in https://github.com/blockscout/blockscout/pull/9469/commits/b00c0b46fd7a9ff0483decdb6222b7cf6f4b14ac

File: apps/explorer/lib/explorer/etherscan/logs.ex (hunk 1 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/etherscan/logs.ex:169]: Not really, I suppose, filter by block range together with created contract address hash should work at these lines https://github.com/blockscout/blockscout/pull/9469/files#diff-5221a499e0ba4352c2b3f0e1755f8d12a1c5d8e4a2877584a3ab04a7fcca7d89R152-R153.

File: apps/explorer/lib/explorer/etherscan/logs.ex [hunk 5]
```diff
@@ -165,6 +161,19 @@ defmodule Explorer.Etherscan.Logs do
...
+        )
+
+      all_transaction_logs_query =
+        all_transaction_logs_query_base
+        |> where([transaction], transaction.created_contract_address_hash == ^address_hash)
+        |> or_where([transaction], ^dynamic)
+
```
Comment (k1rill-fedoseev) [inline apps/explorer/lib/explorer/etherscan/logs.ex:176]: Can we simplify it this way to get rid of the duplicated block range condition? 
```suggestion
      dynamic =
        dynamic(
          [transaction],
          ^Transaction.where_transactions_to_from(address_hash) or
            transaction.created_contract_address_hash == ^address_hash
        )

      all_transaction_logs_query =
        all_transaction_logs_query_base
        |> where([transaction], ^dynamic)
```

Comment (vbaranov) [reply apps/explorer/lib/explorer/etherscan/logs.ex:176]: Yes, I think, this should work. Thanks.

**Review by k1rill-fedoseev — APPROVED** [review]

File: apps/explorer/lib/explorer/etherscan/logs.ex [hunk 6]
```diff
@@ -5,7 +5,7 @@ defmodule Explorer.Etherscan.Logs do
 
   """
 
-  import Ecto.Query, only: [from: 2, limit: 2, where: 3, subquery: 1, order_by: 3, union: 2]
+  import Ecto.Query, only: [dynamic: 2, from: 2, limit: 2, where: 3, or_where: 3, subquery: 1, order_by: 3, union: 2]
```
Comment (nikitosing) [inline apps/explorer/lib/explorer/etherscan/logs.ex:8]: or_where is not needed anymore?
